### PR TITLE
Fix for #11 and #12

### LINKED ISSUE #11: build issue
Hello, can't build the code on 4.13.16
Here is the log:

  CC [M]  /home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.o
/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.c: In function ‘_mali_osk_timer_setcallback’:
/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.c:63:2: error: implicit declaration of function ‘timer_setup’; did you mean ‘ktime_set’? [-Werror=implicit-function-declaration]
  timer_setup(&tim->timer, callback, 0);
  ^~~~~~~~~~~
  ktime_set
cc1: some warnings being treated as errors
scripts/Makefile.build:302: recipe for target '/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.o' failed
make[2]: *** [/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.o] Error 1
Makefile:1512: recipe for target '_module_/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali' failed
make[1]: *** [_module_/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali] Error 2
make[1]: Leaving directory '/home/sergey/Projects/armbian/build/cache/sources/linux-mainline/linux-4.13.y'

### LINKED ISSUE #12: devfreq patch is broken
can't find file to patch at input line 22
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|From 9af33ab01cf84d90fe9a4d354b71dd304bca3010 Mon Sep 17 00:00:00 2001
|From: Maxime Ripard <maxime.ripard@free-electrons.com>
|Date: Thu, 30 Nov 2017 11:45:24 +0100
|Subject: [PATCH] mali: Make devfreq optional
|
|If the USING_DEVFREQ option is set, any device without operating points
|defined in the device tree will be treated as an hard error, making probe
|fail.
|
|Rework the driver a bit to ignore the property missing error, and treat all
|the others as hard errors.
|
|Signed-off-by: Maxime Ripard <maxime.ripard@free-electrons.com>
|---
| r6p2/src/devicedrv/mali/linux/mali_devfreq.c | 7 +++++--
| 1 file changed, 5 insertions(+), 2 deletions(-)
|
|diff --git a/r6p2/src/devicedrv/mali/linux/mali_devfreq.c b/r6p2/src/devicedrv/mali/linux/mali_devfreq.c
|index 0b0ba1481eed..c3a2636e6635 100755
|--- a/r6p2/src/devicedrv/mali/linux/mali_devfreq.c
|+++ b/r6p2/src/devicedrv/mali/linux/mali_devfreq.c
--------------------------
File to patch: 
Skip this patch?

### PR DESCRIPTION
**Pull request by giuliobenetti**
Fix 0014 fixes #12
Fix 0015 fixes #11

### CODE REVIEW

**Comment by giuliobenetti** [comment]
@mripard 
Sorry but I really can't make it work last patch Reverse.
It gives me Unreversed patch detected!
It seems a dependency problem.
Unapply patches using same order than applying can bring to this.
Is there a way to make patch aware of dependency?

**Comment by giuliobenetti** [comment]
@mripard added unapply_patches() for reversing patches from bottom to top.
Now everything seems to be ok.

File: patches/0015-mali-support-building-before-4.14.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ 
++#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
++static inline void timer_setup(struct timer_list *timer,
++			       void (*callback)(struct timer_list *),
++			       unsigned int flags)
++{
++	__setup_timer(timer, (TIMER_FUNC_TYPE)callback,
++		      (TIMER_DATA_TYPE)timer, flags);
++}
++#endif
```
Comment (mripard) [inline patches/0015-mali-support-building-before-4.14.patch]: That one shouldn't be part of a different patch, but fix the previous one.

File: patches/0013-mali-support-building-against-4.15.patch [hunk 2]
```diff
@@ -108,7 +108,7 @@ Signed-off-by: John Stultz <john.stultz@linaro.org>
  	MALI_DEBUG_ASSERT_POINTER(tim);
 -	tim->timer.data = (unsigned long)data;
 -	tim->timer.function = (timer_timeout_function_t)callback;
-+	timer_setup(&tim->timer, callback, 0);
++	__setup_timer(&tim->timer, callback, &tim->timer, 0);
  }
```
Comment (mripard) [inline patches/0013-mali-support-building-against-4.15.patch]: And using your previous patch, you dont need that one anymore

**Comment by giuliobenetti** [comment]
@mripard 
Fix 0014 fixes #12
Fix 0013 fixes #11
Added also unapply_patches to reverse patches order to reverse patches

File: patches/0013-mali-support-building-against-4.15.patch [hunk 3]
```diff
@@ -1,17 +1,22 @@
-From 5bc1f96c6e9b8927e3020fdfe58b2d99fa42bcb3 Mon Sep 17 00:00:00 2001
-From: John Stultz <john.stultz@linaro.org>
-Date: Thu, 16 Nov 2017 13:42:23 -0800
-Subject: MALI: v4.15-rc build fix
+From 1fd429daef6dc5f4b2efbf032caacb473f027364 Mon Sep 17 00:00:00 2001
+From: Giulio Benetti <giulio.benetti@micronovasrl.com>
+Date: Tue, 5 Dec 2017 22:32:31 +0100
+Subject: [PATCH 1/1] mali support building against 4.15-rc
+
+Used new timer api provided by <linux/timer.h>
```
Comment (mripard) [inline patches/0013-mali-support-building-against-4.15.patch]: You're not the primary author of this patch. You shouldn't change that part.

File: patches/0013-mali-support-building-against-4.15.patch [hunk 4]
```diff
@@ -1,17 +1,22 @@
...
- 5 files changed, 15 insertions(+), 13 deletions(-)
+ src/devicedrv/mali/common/mali_control_timer.c  |  2 +-
+ src/devicedrv/mali/common/mali_group.c          |  8 +++++---
+ src/devicedrv/mali/common/mali_osk_types.h      |  8 +++++++-
+ src/devicedrv/mali/linux/mali_memory_os_alloc.c |  2 +-
+ src/devicedrv/mali/linux/mali_osk_timers.c      | 16 ++++++++++------
+ 5 files changed, 24 insertions(+), 12 deletions(-)
 
+diff --git a/src/devicedrv/mali/common/mali_control_timer.c b/src/devicedrv/mali/common/mali_control_timer.c
+index fc6ceb4..57f3c3a 100755
```
Comment (mripard) [inline patches/0013-mali-support-building-against-4.15.patch]: Or that one.

File: patches/0013-mali-support-building-against-4.15.patch [hunk 5]
```diff
@@ -23,9 +28,11 @@ Signed-off-by: John Stultz <john.stultz@linaro.org>
  {
  	if (mali_utilization_enabled()) {
  		struct mali_gpu_utilization_data *util_data = NULL;
+diff --git a/src/devicedrv/mali/common/mali_group.c b/src/devicedrv/mali/common/mali_group.c
+index 9e31802..b9a0d0e 100755
 --- a/src/devicedrv/mali/common/mali_group.c
 +++ b/src/devicedrv/mali/common/mali_group.c
-@@ -44,7 +44,7 @@ int mali_max_job_runtime = MALI_MAX_JOB_
+@@ -44,7 +44,7 @@ int mali_max_job_runtime = MALI_MAX_JOB_RUNTIME_DEFAULT;
```
Comment (mripard) [inline patches/0013-mali-support-building-against-4.15.patch]: I guess you get the idea... :)

File: patches/0013-mali-support-building-against-4.15.patch [hunk 6]
```diff
@@ -1,17 +1,22 @@
...
- src/devicedrv/mali/common/mali_osk_types.h      | 8 +++++++-
- src/devicedrv/mali/linux/mali_memory_os_alloc.c | 2 +-
- src/devicedrv/mali/linux/mali_osk_timers.c      | 8 +-------
- 5 files changed, 15 insertions(+), 13 deletions(-)
+ src/devicedrv/mali/common/mali_control_timer.c  |  2 +-
+ src/devicedrv/mali/common/mali_group.c          |  8 +++++---
+ src/devicedrv/mali/common/mali_osk_types.h      |  8 +++++++-
+ src/devicedrv/mali/linux/mali_memory_os_alloc.c |  2 +-
+ src/devicedrv/mali/linux/mali_osk_timers.c      | 16 ++++++++++------
+ 5 files changed, 24 insertions(+), 12 deletions(-)
```
Comment (mripard) [inline patches/0013-mali-support-building-against-4.15.patch]: Or that part.

**Comment by giuliobenetti** [comment]
@mripard now everything should be ok!
Let me know, thank you

File: patches/0013-mali-support-building-against-4.15.patch (hunk 6 above)
Comment (giuliobenetti) [reply patches/0013-mali-support-building-against-4.15.patch]: @mripard This depends on the fact I've added timer_setup function in case kernel_version < 4.14
